csharp_dotnetcore/17.Proactive-Messages/ProactiveBot.cs

```diff
@@ -0,0 +1,208 @@
+
+                        break;
+
+                    case "show":
+                    case "show jobs":
+
+                        // Display information for all jobs in the log.
+                        if (jobLog.Count > 0)
+                        {
+                            await turnContext.SendActivityAsync(
+                                "| Job number &nbsp; | Conversation ID &nbsp; | Completed |<br>" +
+                                "| :--- | :---: | :---: |<br>" +
+                                string.Join("<br>", jobLog.Values.Select(j =>
+                                    $"| {j.TimeStamp} &nbsp; | {j.Conversation.Conversation.Id} &nbsp; | {j.Completed} |")));
+                        }
+                        else
+                        {
+                            await turnContext.SendActivityAsync("The job log is empty.");
+                        }
+
+                        break;
+
+                    default:
+
+                        // Check whether this is simulating a job completed event.
+                        string[] parts = text?.Split(' ', StringSplitOptions.RemoveEmptyEntries);
+                        if (parts != null && parts.Length is 2
+                            && parts[0].Equals("done", StringComparison.InvariantCultureIgnoreCase)
+                            && long.TryParse(parts[1], out long jobNumber))
+                        {
+                            if (!jobLog.TryGetValue(jobNumber, out JobLog.JobData jobInfo))
+                            {
+                                await turnContext.SendActivityAsync($"The log does not contain a job {jobInfo.TimeStamp}.");
+                            }
+                            else if (jobInfo.Completed)
+                            {
+                                await turnContext.SendActivityAsync($"Job {jobInfo.TimeStamp} is already complete.");
+                            }
+                            else
+                            {
+                                await turnContext.SendActivityAsync($"Completing job {jobInfo.TimeStamp}.");
+
+                                await CompleteJobAsync(turnContext.Adapter, AppId, jobInfo);
+                            }
+                        }
+
+                        break;
+                }
+
+                if (!turnContext.Responded)
+                {
+                    await turnContext.SendActivityAsync(
+                        "Type `run` or `run job` to start a new job.<br>" +
+                        "Type `show` or `show jobs` to display the job log.<br>" +
+                        "Type `done <jobNumber>` to complete a job.");
+                }
+            }
+        }
+
+        /// <summary>Handles non-message activities.</summary>
+        /// <param name="turnContext">The context object for this turn.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
```

Comment:
private method, doesn't need these comments. Consider putting them as normal comments in the code?

Reply:
The /// comments provide intellisense feedback in VS. I often include it for the dev's convenience, if they're actually following along in VS. It isn't necessary, however.